File: fs/pkg/inode/store/volumeinodestore.go
package store

import (
	"fmt"

	"github.com/weberc2/mono/fs/pkg/encode"
	"github.com/weberc2/mono/fs/pkg/io"
	. "github.com/weberc2/mono/fs/pkg/types"
)

type VolumeInodeStore struct {
	volume io.Volume
}

func NewVolumeInodeStore(volume io.Volume) VolumeInodeStore {
	return VolumeInodeStore{volume}
}

func (store VolumeInodeStore) Put(inode *Inode) error {
	buf := new([InodeSize]byte)
	encode.EncodeInode(inode, buf)
	offset := Byte(inode.Ino) * InodeSize
	if err := store.volume.WriteAt(offset, buf[:]); err != nil {
		return fmt.Errorf(
			"writing inode `%d` to volume at offset `%d`: %w",
			inode.Ino,
			offset,
			err,
		)
	}
	return nil
}

func (store VolumeInodeStore) Get(ino Ino, output *Inode) error {
	buf := new([InodeSize]byte)
	offset := Byte(ino) * InodeSize
	if err := store.volume.ReadAt(offset, buf[:]); err != nil {
		return fmt.Errorf(
			"reading inode `%d` from volume at offset `%d`: %w",
			ino,
			offset,
			err,
		)
	}
	encode.DecodeInode(output, buf)
	output.Ino = ino
	return nil
}

File: rustfs/src/volumeinodestore.rs
use crate::byte::Byte;
use crate::inode::{Ino, Inode};
use crate::inodestore::InodeStore;
use crate::io::{ReadAt, WriteAt};
use crate::write::{read_inode, write_inode};
use std::io::Result;

pub struct VolumeInodeStore<V> {
    volume: V,
}

impl<V> VolumeInodeStore<V> {
    pub fn new(volume: V) -> VolumeInodeStore<V> {
        VolumeInodeStore { volume }
    }
}

impl<V: ReadAt + WriteAt> InodeStore for VolumeInodeStore<V> {
    fn get(&mut self, ino: Ino) -> Result<Option<Inode>> {
        Ok(Some(read_inode(&mut self.volume, ino)?))
    }

    fn put(&mut self, inode: &Inode) -> Result<()> {
        write_inode(&mut self.volume, inode)
    }
}
